Require 'basic_debug' priv to view gameplay-relevant debug info (2nd try), require 'debug' priv to view wireframe

Wuzzy2 opened the pull request:
Fixes #7245.

This is a revival of PR #8951, but with these changes:

* Profiler screen is no longer touched (it was a huge mistake to add this completely unrelated feature in the PR)
* Added `basic_debug` priv

## What this PR does

* Hide the position, yaw, pitch, mapseed and pointed thing if player does not have the `basic_debug` priv. In practice, the 2nd line of the debug text at the top will disappear and also the entity info when you point an entity. Additionally, the block bounds can't be used as well
* Add `basic_debug` priv to unlock various debug info that could be considered a gameplay advantage if they were accessible by default. NOT granted by default
* With the `debug` priv (which is not new) you also unlock the wireframe

## Justification

The F5 debug screen contains various "spoilers", e.g. info that can give you a gameplay advantage for free: position, pointed node, pointed entity (including health and armor groups). This completely ruins mods like `orienteering` and `compassgps` in which you can craft items that expose your position and more. The mods get ruined because there's no point in crafting a compass if you can get the yaw/position/whatever “for free” with a single key press, and it's unfair to force players to resist the temptation. Players complained, as evidenced by #7245, which also has core dev support.

This PR gets rid of those spoilers unless you have `basic_debug` priv. `orienteering`-style mods can now be created with peace-of-mind.

## To do

Nothing, for now.

## How to test

* Join any world
* Press F5 while you have one of the 4 possible combinations of “you have / have not `debug` priv” and “you have / have not `basic_debug` priv”.

## Expected questions and counter-arguments

1. “This PR is pointless because it's the player's fault for pressing F5.” This argument is bad because it's not good to force players to resist the temptation to enable a cheat that is only 1 keypress away. You could also press F5 by accident, and you can't “unsee” the coordinates. This is not good if a key element of a game is to “earn” the privilege to learn your coordinates through gameplay
2. “Why not using HUD flags?”. HUD flags are controlled by mods. This implies that mods can screw around with the debug info at will. All it takes is one poorly-coded mod to strip away the extended debug info from the admin, which is not fun to track down. Privileges, by comparison, are much more reliable and predictable and there are no ugly surprises.
3. “But how do I get the old behavior back?”: Type `/grantme basic_debug`.
4. “My game/mod/server depends on the pos/angle/pointed thing be visible in debug screen, this PR ruins everything!”. It was never a good idea for a mod to depend on anything written in the debug screen. As the name suggests, its for debugging primarily. The fact that it's currently being abused heavily for gameplay as well is just an excuse for bad practice. Instead of depending on the debug screen, add a mod that exposes the desired info itself, there are countless mods that do this which also can do this with a much nicer HUD than the debug screen. “Knowing the coordinates” is a gameplay advantage and should be considered as such. But you don't need to install any mods, `/grantme basic_debug` still works.

style-nine commented:
I have to admit that I don't really see why this doesn't use HUD flags. If an admin is surprised by what a mod is doing, then doesn't that mean that either the mods are too complicated to police properly or that the admins aren't as alert as they should be?

Wuzzy2 commented:
Mods should have NO business in touching the debug info at all. Debug info must work without mods, it's a very essential feature. Allowing mods to screw around with this just makes things unneccessarily complicated. I chose privileges precisely because they are straight-forward, work out of the box and are reliable, no matter what crazy tower of mods you have installed. Also, you failed to provide any argument *for* HUD flags.

style-nine commented:
It may be debug info, but it uses the HUD to present it. Using HUD flags is consistent with the game as it's already implemented. Treating the mods like an unknown and a danger is in itself a dangerous habit to reinforce that is totally unecessary. Mods are entirely within the control of the server admins. Mods themselves can regulate other mods. You can always implement a master mod that governs the others, and privs can always be circumvented with unknown mods, so I don't see how privs is a fix where mods are not.

LoneWolfHT commented:
Unless support was added for mods using debug info to display their own values I see no reason for mods to control this.
If you want all your players to see their coordinates on their HUD it's not hard to make a mod to do that, which is probably what we should have been doing in the first place

paramat commented on builtin/game/privileges.lua:
```diff
@@ -94,6 +94,10 @@ core.register_privilege("debug", {
 	give_to_singleplayer = false,
```
true

paramat commented:
I like how this moves the wireframe viewing priv to a more suitably named priv 'wireframe'. This then frees up the 'debug' priv for another function.

As requested by 2 core devs in the first PR, the default for players of this new 'debug' priv for viewing position, yaw, pitch, mapseed and pointed thing should be true, currently it is false.
Other than that i have no objections.

paramat commented on builtin/game/privileges.lua:
```diff
@@ -94,6 +94,10 @@ core.register_privilege("debug", {
 	give_to_admin = true,
 })
+core.register_privilege("wireframe", {
+	description = "Can enable wireframe",
```
'Can enable' is awkward English and has a double meaning, better is "Allows viewing wireframe mode".

paramat replied:
Oh ... just seen many priv descriptions use "Can ...", that is bad =)

paramat replied:
So, i am not too bothered about this being improved, up to you. We could improve all at once later.

paramat replied:
Comment now invalid as privs need changing.

Wuzzy2 commented:
Sadly, granting `debug` to singleplayer by default means you can't get rid of it in singleplayer, as all privs granted by default are irrevokable. This would defeat the very purpose of this PR. This is unacceptable!

Wuzzy2 commented:
Also, see the FAQ above: Just type in `/grantme debug` and you'll get back the old behavior. Or maybe even add it into your default privs in `minetest.conf`.

paramat commented:
> Sadly, granting debug to singleplayer by default means you can't get rid of it in singleplayer, as all privs granted by default are irrevokable. This would defeat the very purpose of this PR.

Oh hmm ... i did not know this, sorry, removing my -1.
However, the objection from core devs remains, so this seems to be a dilemma.

style-nine commented:
So, the inability to revoke a priv in singleplayer if it's default would be a bug, don't you think?

paramat commented:
I agree, we should consider changing that.

rubenwardy commented:
> So, the inability to revoke a priv in singleplayer if it's default would be a bug, don't you think?

You can revoke default privileges in singleplayer.

There's a feature called `give_to_singleplayer` which means that singleplayer always has certain privileges such as interact, hence why they can't be removed

rubenwardy commented:
The changing of the meaning of `debug` will cause older clients to be able to use the wireframe if they are given that privilege

paramat commented:
Thanks.
Yes, looks like 'debug' has to remain for wireframe, and a new priv name added.

Wuzzy2 commented:
I don't understand. How can the player possibly revoke their default privs?

rubenwardy commented:
minetest.conf:

```
default_privs = fly,fast,noclip
```

in game: `/revoke singleplayer all`

Wuzzy2 commented:
Sigh. OK, maybe the irrevokable privs can be fixed in a different PR then.

Just to be clear: With this change, `debug` will become an irrevokable priv in singleplayer.

paramat commented on builtin/game/privileges.lua:
```diff
@@ -91,9 +91,13 @@ core.register_privilege("rollback", {
 	description = "Allows enabling various debug options that may affect gameplay",
-	give_to_singleplayer = false,
+	give_to_singleplayer = true,
 	give_to_admin = true,
```
Due to
```
		if def.give_to_singleplayer == nil then
			def.give_to_singleplayer = true
		end
		if def.give_to_admin == nil then
			def.give_to_admin = def.give_to_singleplayer
		end
```
Both lines can be removed.

paramat replied:
Done.

paramat commented:
> The changing of the meaning of debug will cause older clients to be able to use the wireframe if they are given that privilege

Unfortunately 'debug' has to remain for wireframe, and a new priv name added for position/direction/seed/pointed.

Wuzzy2 commented:
@paramat: I have removed the two lines.
Sadly, I do not know how to proceed.

paramat commented:
Seems to me the following is necessary:

The 'debug' priv must continue to control wireframe, for the reason explained above.
The description of 'debug' must be changed to be clear it is for wireframe mode only.
We need to choose a new name for your new priv that controls 'basic debug text'.

Wuzzy2 commented:
Using the `debug` name exclusively for wireframe is a little confusing name, however. Just teach server operators to update their privs.

I am not really happy with this proposal. :-(

Is there no better solution for this?

paramat commented:
Yeah it is messy and unfortunate, but as 'debug' is granted by default and usually granted, all older clients will then be able to use wireframe mode, causing cheating on servers.
Even if we try to inform, many servers will simply not make the necessary changes, the risk seems too big.

Wuzzy2 commented:
Oh crap. Auto-granting wireframe to everyone is not good. Yeah, you got a point.

Wuzzy2 commented:
New changes!

* `debug` priv is no for wireframe only (old behavior) (renamed from `wireframe`)
* New priv: `basic_debug` for enabling displaying gameplay-relevant debug info text (renamed from `debug`)
* The `wireframe` priv is gone
* Both privs are `give_to_singleplayer = false`

paramat commented:
The 'debug' priv (wireframe) should also have `give_to_admin = true,`, as before.
The 'basic_debug' priv should be granted by default, as discussed earlier. Priv name seems fine to me.

Wuzzy2 commented:
Hopefully I did it right this time.

paramat commented on builtin/game/privileges.lua:
```diff
@@ -89,8 +89,12 @@ core.register_privilege("rollback", {
 })
+core.register_privilege("basic_debug", {
+	description = "Can view more debug info that might give a gameplay advantage",
+	give_to_singleplayer = true,
```
Line unnecessary due to default behaviour.

paramat commented:
Fine, apart from an unnecessary line.

Wuzzy2 commented:
Done!

BluebirdGreycoat commented:
Status of this PR? My server has a custom coordinate system (offset from the real coords, with multiple realms), and being able to see the debug coordinates often causes confusion for new players, plus it tends to break the immersion. Would be nice to be able to disable it and ensure everyone uses the same custom coordinate system.

Wuzzy2 commented:
Status: Waiting for (re-)review.

Warr1024 commented:
This is one of the top engine issues that's been an obstacle to NodeCore, with no workaround possible, and it has a notable impact on both singleplayer and multiplayer, so I'm looking forward very much to seeing it merged.

Wuzzy2 commented:
Bump after 5 months.

appgurueu commented:
`basic_debug` and `debug` privs should be granted by default. This can't be done using current settings, as those are already set on older servers and there should be no breaking changes.

:+1: for the feature, :-1: for breaking compat

Wuzzy2 commented:
How does this break compat?
I disagree that they should be granted by default.
To admin: Maybe, but probably not. To singleplayer: Definitely not. Why? Because most players are … well … players, and not devs. Therefore, disabling it is the more reasonable default.

But even if I would agree, the problem is: The only way to currently grant anything by default (via the engine) is by setting (or not setting) one of the `give_to_singleplayer` or `give_to_admin`. The problem here is that ALL privs granted by default are **irrevokable**. This would be a serious annoyance and defeats the purpose of this PR. Especially for singleplayer

Besides, there is a setting somewhere in which you can already change your initial privileges. I think this covers your use case just fine, so no changes to the code are neccessary after all. :P

appgurueu commented:
No. That setting does not apply to already joined players and would need to be changed by admins. Many players rely on debug for coordinates.

BluebirdGreycoat commented:
Bump. I really want to be able to turn debug coordinates off.

GreenXenith commented:
Agreed, this is a very necessary step towards giving mods/games better control.
Are there any issues remaining to solve?

appgurueu commented:
I <small>(now)</small> support this. Wireframe etc. is not gameplay relevant so it makes sense to disable it by default. It would be nice to keep backwards compatibility for basic_debug, having it default to true and requiring it to be explicitly disabled, but in order to achieve this, dirty hacks would be needed, such as having a basic_debug_disabled priv, which definitely is a worse solution.

Requiring admins to overthink whether they want their users to be able to have permanent access to coords is therefore justified.

hecktest commented:
minetest_game can grant players basic_debug on join, everyone else should be forced to make a decision whether they want the spoiler hud back. I sure won't miss it.

Problem is, any punk with a debugger can obtain this info anyway as long as it is sent to the client.

appgurueu commented:
> Problem is, any punk with a debugger can obtain this info anyway as long as it is sent to the client.

Punks without debuggers can too. This unfortunately is nontrivial to fix.

1. Not sending absolute positions to the client would greatly complicate things.
2. Not sending obstructed chunks & objects within them would require significant server effort & nontrivial visibility tests.

lhofhansl commented:
In the end I would have been happier with a "debug" and "wireframe" privs. But since that is not apparently possible this is good.

I do *not* think that these privs should be granted to anyone by default. If you're an "admin" and cannot figure our to grant these privs to yourself then ... 

So +1 on this in concept. I'll try it out today.

lhofhansl commented:
So if I'm the admin of the server should I be able to revoke the privs or not? Right now I get them by default and I cannot revoke them.

Wuzzy2 commented:
Okay, I should clarify: The point of this PR is !!!NOT!!! to implement an effective anti-cheat. I am aware that hiding the coordinates is not enough if we want to seriously hide this info. I

It's more about letting games give control about whether they want to give the player coordinates for free (it's only 1 keypress away, after all) or if they have to earn it (e.g. by crafting a tool or something).
Right now, F5 acts like an "1-click cheat" for all games that do not want to give you the coordinates for free.

lhofhansl commented:
@Wuzzy2 totally agree. I agree with what you want to do, and it's a shame it takes almost a year to do that.

When I tested it, though, both admins and singleplayer still have basic_debug by default (and that cannot be revoked), and I thought that is what you wanted to avoid.

appgurueu commented:
I would like to request an addition: a priv `disable_hud` for F1 (disabling the HUD), as that can be used to evade gameplay effects such as flash grenades. Disabling particles should also be controllable by the server.

Wuzzy2 commented:
> I would like to request an addition: a priv disable_hud for F1 (disabling the HUD), as that can be used to evade gameplay effects such as flash grenades. Disabling particles should also be controllable by the server.

This is unrelated to this PR. Besides, I think it would be better to be able to force certain HUD elements to be unhidable, rather than disallowing to disable to hide the whole HUD.

Wuzzy2 commented:
This PR has become old and dusty, so I have updated and rebased it to 5.4.0-dev.

I have also removed giving the privileges to anyone by default, as requested before.

daroberts00 commented:
As my game/server nears completion, personally it frustrates me greatly the extra work I have to put in to quests etc in a survival/semi-RPG server just because the coordinates are there. Wuzzy has a mod that exposes these values if a player crafts the items, but then if the server owner wants the game to be even more "hardcore" - there's always the sun! It should be up to the game designer how their game works, and the server owner what info is provided to the player.

celeron55 commented:
This was and is still a good idea. I believe rubenwardy will hit the merge button soon enough if this still works.

rubenwardy commented:
If you grant/revoke `basic_debug` when the debug interface is open, nothing changes. It should probably trigger an update when that privilege is modified

rubenwardy commented:
Okay, to summarise that way too long IRC discussion: the priv is fine and so is removing it by default too.

The only changes required are:

1. As mentioned above, an update check to fix the bug with priv changes
2. A protocol version check to not require the priv on 5.4 or lower servers

Wuzzy2 commented:
This PR just got rebased ... again.

There are 2 new commits.

I have finally fixed the issue that the debug screen does not update on `basic_debug` priv change. The debug screen will now update instantly as required.

Also, I have applied a new restricion to the "mapblock bounds" feature with F8. This key does not work unless you have the `basic_debug` privilege. The reason is that knowing where the mapblock boundaries are might also be considered a gameplay advantage (even if a minor). Also, it's technically also a debug thing. This is me just following the spirit of #7245.

Wuzzy2 commented:
Protocol version check added but not tested by me. **This PR now bumps the protocol to 40!**

Thanks to Warr1024.

rubenwardy commented on src/network/clientpackethandler.cpp:
```diff
@@ -896,6 +896,10 @@ void Client::handleCommand_Privileges(NetworkPacket* pkt)
 		infostream << priv << " ";
 	}
+
+	if(m_proto_ver < 40)
```
missing space, maybe a comment would be good too

rubenwardy commented:
Note: I tested everything except the protocol version bump

Warr1024 commented:
I have tested this and it works as expected, and the code looks fine to me, though I don't feel confident enough to give it the final merging approval.

- 5.5 client on 5.5 server: works, selected info is hidden/unavailable, granting and revoking the priv behaves as expected.
- 5.5 client on 5.4 server: works, selected info is still shown as per the existing 5.4 behavior.
- 5.4 client on 5.5 server: not applicable as the changes only affect the 5.5+ client.

celeron55 commented:
I see this was tested to work and the code looks fine to me.

appgurueu commented:
Yay

sharedrory commented:
Not cool. Soon everything with require a priv. Respawning, looking around, jumping, clicking, audio, joining a serv-

Wuzzy2 commented:
@sharedrory: Slippery slope fallacy.

appgurueu commented:
> @sharedrory: Slippery slope fallacy.

Slippery slope? I'd rather call this fallacy "taking it to the extremes".

These debug features should never have been included in non-debug clients in the first place.

sharedrory commented:
> @sharedrory: Slippery slope fallacy.

I don't think you've considered the number of servers that DON'T use those mods. Now we have to go bug the server owners to give us the priv. The fact that you don't get it in singleplayer means you now have to type out yet another 2 commands to get it unless you want to grant all at the same time.

> These debug features should never have been included in non-debug clients in the first place.

Then atleast have mtg add a hud element with the coords? Perhaps a system like the minimap where you need the mapping kit. Speaking of the mapping kit, it reveals the direction you are facing (north, south, ...).

appgurueu commented:
> I don't think you've considered the number of servers that DON'T use those mods. Now we have to go bug the server owners to give us the priv.

You should not bug them to give you the priv. You should bug them to install a mod which provides the same essential functionality (coords, I assume).

sharedrory commented:
> You should **not bug** them to give you the priv. You should **bug** them to install a mod which provides **the same** **_essential_** functionality (coords, I assume).

Why don't we make the engine available behind a priv. And then let a mod provide the same essential functionality.

sfan5 commented:
If it's widely considered not to be a problem then Minetest Game could also grant this priv by default. Other games would still have the freedom not to do this and users will be happy that they don't have to bother server admins for this.

appgurueu commented:
> If it's widely considered not to be a problem then Minetest Game could also grant this priv by default. Other games would still have the freedom not to do this and users will be happy that they don't have to bother server admins for this.

The only information that really is must be displayed are the coordinates, which are relied on for navigation. The other clutter isn't needed.
